### PR TITLE
Add UUID From RequestContext

### LINKED ISSUE #2001: Add UUID to request logging
Currently logs look like

```
.c.l.LoggingClient                  : Request: {startTime=2019-08-17T12:39:19.994Z(1566045559994257), length=4970KiB(5089733B), duration=6198?s(6198300ns), scheme=none+h1c, headers=[]}
zipkin                      | 2019-08-17 12:39:20.117  WARN 1 --- [orker-epoll-2-4] c.l.a.c.l.LoggingClient                  : Response: {startTime=2019-08-17T12:39:20.117Z(1566045560117742), length=0B, duration=0ns, cause=com.linecorp.armeria.client.ResponseTimeoutException, headers=[]}
```

It would be useful, and probably simple, to assign a UUID to every request to add to the log to be able to associate requests and responses. Many users will be setting trace IDs that would perform similar utility, but not all of them and I think there's very little cost even if it ends up performing double-purpose.

### PR DESCRIPTION
**Pull request by yanghun0070**
Fix #2001 
Add UUID From RequestContext

### CODE REVIEW

**Comment by yanghun0070** [comment]
It is easier to request a new pull request than to modify the source code.
https://github.com/line/armeria/pull/2174

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java [hunk 1]
```diff
@@ -158,8 +158,8 @@ private DefaultServiceRequestContext(
             InetAddress clientAddress, boolean requestStartTimeSet, long requestStartTimeNanos,
             long requestStartTimeMicros) {
 
-        super(meterRegistry, sessionProtocol,
-              requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
+        super(meterRegistry, sessionProtocol, null,
+              requireNonNull(routingContext).method(), routingContext.path(),
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]:
```suggestion
              requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
```

File: core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java [hunk 2]
```diff
@@ -158,8 +158,8 @@ private DefaultServiceRequestContext(
             InetAddress clientAddress, boolean requestStartTimeSet, long requestStartTimeNanos,
             long requestStartTimeMicros) {
 
-        super(meterRegistry, sessionProtocol,
-              requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
+        super(meterRegistry, sessionProtocol, null,
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java]: I don’t think we can put `null` here.

**Comment by minwoox** [comment]
Could you add a method in `ArmeriaServerHttpRequest`?

```java
@Override
protected String initId() {
    return ctx.uuid().toString();
}
```

**Comment by yanghun0070** [comment]
@minwoox 
not use uuid in ServiceRequestContext.

File: core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java [hunk 3]
```diff
@@ -44,6 +45,8 @@
     private final MeterRegistry meterRegistry;
     private final DefaultAttributeMap attrs = new DefaultAttributeMap();
     private final SessionProtocol sessionProtocol;
+    @Nullable
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java]: Could we remove this? :-)

**Comment by minwoox** [comment]
> not use uuid in ServiceRequestContext.

Is there any reason that you don't want to add `UUID` in `ServiceRequestContext`?

**Comment by yanghun0070** [comment]
I think. :)
`ServiceRequestContext` is related to `RoutingContext`.
`RoutingContext` was used to search for domain mapped services.

So unnecessary and i removed the `RoutingContext` uuid in review.

**Comment by minwoox** [comment]
`RoutingContext` holds the parameters which are required to find service available to handle the request from its Javadoc.
So it's only used to find a service. That's why I asked you to remove the `UUID` from it.
On the other hand, `ServiceRequestContext` is a context that has information about a request and response on the server-side. From #2001, we need a UUID to associate requests and responses using `requestLog`.
If we only add the `UUID` to `ClientRequestContext`, we could not achieve that behavior on the server-side. So we should add it to `ServiceRequestContext` too.
Could you do that?

**Comment by yanghun0070** [comment]
Thank you for a detail description. :)
I'll work on it today.

**Comment by yanghun0070** [comment]
Thank you for a detail description

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 4]
```diff
@@ -64,7 +66,7 @@
 
     private static final AtomicReferenceFieldUpdater<DefaultClientRequestContext, HttpHeaders>
             additionalRequestHeadersUpdater = AtomicReferenceFieldUpdater.newUpdater(
-                    DefaultClientRequestContext.class, HttpHeaders.class, "additionalRequestHeaders");
+            DefaultClientRequestContext.class, HttpHeaders.class, "additionalRequestHeaders");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java]: Revert

File: core/src/main/java/com/linecorp/armeria/client/UserClient.java [hunk 5]
```diff
@@ -151,10 +152,12 @@ protected final O execute(@Nullable EventLoop eventLoop, Endpoint endpoint,
...
         if (eventLoop == null) {
             ctx = new DefaultClientRequestContext(factory(), meterRegistry, sessionProtocol,
-                                                  method, path, query, fragment, options(), httpReq, rpcReq);
+                                                  UUID.randomUUID(), method, path, query, fragment, options(),
+                                                  httpReq, rpcReq);
         } else {
             ctx = new DefaultClientRequestContext(eventLoop, meterRegistry, sessionProtocol,
-                                                  method, path, query, fragment, options(), httpReq, rpcReq);
+                                                  UUID.randomUUID(), method, path, query, fragment, options(),
+                                                  httpReq, rpcReq);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/UserClient.java]: `UUID.randomUUID()` could be extracted into a common local variable?

File: core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java [hunk 6]
```diff
@@ -98,7 +101,6 @@ protected AbstractRequestContextBuilder(boolean server, HttpRequest req) {
         this.req = requireNonNull(req, "req");
         rpcReq = null;
         sessionProtocol = SessionProtocol.H2C;
-
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java]: Revert

File: core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java [hunk 7]
```diff
@@ -67,16 +69,18 @@
...
      */
     protected NonWrappingRequestContext(
             MeterRegistry meterRegistry, SessionProtocol sessionProtocol,
-            HttpMethod method, String path, @Nullable String query,
+            UUID uuid, HttpMethod method, String path, @Nullable String query,
             @Nullable HttpRequest req, @Nullable RpcRequest rpcReq) {
 
         this.meterRegistry = requireNonNull(meterRegistry, "meterRegistry");
         this.sessionProtocol = requireNonNull(sessionProtocol, "sessionProtocol");
+        this.uuid = uuid;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java]: `requireNonNull(uuid, "uuid")`

File: core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java [hunk 8]
```diff
@@ -274,4 +275,18 @@ public void toStringResponseBuilderCapacity() {
...
+        final RequestHeaders reqHeaders =
+                RequestHeaders.of(HttpMethod.POST, "/armeria/uuid",
+                                  HttpHeaderNames.CONTENT_LENGTH, VERY_LONG_STRING.length());
+        final HttpRequest req = HttpRequest.of(
+                AggregatedHttpRequest.of(reqHeaders, HttpData.ofUtf8(VERY_LONG_STRING)));
+        final ClientRequestContext cctx = ClientRequestContextBuilder.of(req).build();
+        assertThat(cctx.log().uuid()).isNotNull();
+
+        final ServiceRequestContext sctx = ServiceRequestContext.of(req);
+        assertThat(sctx.log().uuid()).isNotNull();
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/common/logging/DefaultRequestLogTest.java]: Could ensure `ctx.log().uuid()` is equal to `ctx.uuid()` as well as checking non-nullness.

File: spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java [hunk 9]
```diff
@@ -92,6 +92,11 @@ private static HttpHeaders fromArmeriaHttpHeaders(com.linecorp.armeria.common.Ht
         return cookies;
     }
 
+    @Override
+    protected String initId() {
+        return ctx.uuid().toString();
```
Comment (trustin) [inline spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java]: Nice catch :+1:

File: spring/boot-webflux-autoconfigure/src/test/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequestTest.java [hunk 10]
```diff
@@ -67,6 +67,7 @@ void readBodyStream() {
 
         final ServiceRequestContext ctx = newRequestContext(httpRequest);
         final ArmeriaServerHttpRequest req = request(ctx);
+        assertThat(req.initId()).isNotEmpty();
```
Comment (trustin) [inline spring/boot-webflux-autoconfigure/src/test/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequestTest.java]: Could ensure `req.init()` is equal to `ctx.uuid().toString()`.

**Comment by trustin** [comment]
@yanghun0070 Could you also make sure the CLA is signed correctly and your commit has a correct e-mail address?

**Comment by yanghun0070** [comment]
I have signed the CLA now.

File: spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java (hunk 9 above)
Comment (yanghun0070) [reply spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpRequest.java]: This was corrected with Minwoo help. :)  👍

File: core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java [hunk 11]
```diff
@@ -433,6 +435,24 @@ protected final String path() {
         return path;
     }
 
+    /**
+     * Sets the {@link UUID} of the request. If not set, it is auto-generated with {@link UUID}
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java]: it is auto-generated with ... -> `a random {@link UUID} is generated with {@link UUID#randomUUID()}.`

File: core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java (hunk 6 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java]: Seems this is not reverted?

File: core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java (hunk 7 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java]: Seems not fixed?

Comment (yanghun0070) [reply core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java]: Sorry my mistake. Corrected~!

File: core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java (hunk 6 above)
Comment (yanghun0070) [reply core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java]: Sorry my mistake. 
Corrected~!

**Pull request merged**